test_spack.py

```diff
@@ -0,0 +1,432 @@
+    while True:
+        old_size = len(reachers)
+        # add elements up the DAG
+        reachers |= {parent for parent, children in DAG.items() if any(child in reachers for child in children)}
+        new_size = len(reachers)
+        if old_size == new_size:
+            return reachers
+
+def Has_cycle(directed_graph: map) -> bool:
+    return False # TODO: Implement!
+
+
+class DAG_Algorithm_Test(unittest.TestCase):
+
+    def test_direction(self):
+        # a -> b -> c
+        DAG = {'a': {'b'}, 'b': {'c'}}
+        reachers = Self_and_up({'b'}, DAG)
+        self.assertTrue('a' in reachers)
+
+    def test_transitivity(self):
+        # a -> b -> c -> d
+        DAG = {'a': {'b'}, 'b': {'c'}, 'c': {'d'}}
+        reachers = Self_and_up({'c'}, DAG)
+        self.assertEqual(reachers, {'a','b','c'})
+
+    def test_diamon(self):
+        # a -> b+c -> d
+        DAG = {'a': {'b','c'}, 'b': {'d'}, 'c': {'d'}}
+        b_reachers = Self_and_up({'b'}, DAG)
+        c_reachers = Self_and_up({'c'}, DAG)
+        d_reachers = Self_and_up({'d'}, DAG)
+        self.assertEqual(b_reachers, {'a','b'})
+        self.assertEqual(c_reachers, {'a','c'})
+        self.assertEqual(d_reachers, {'a','b','c','d'})
+
+    # def test_cycle(self): TODO: Uncomment once Has_cycle is implemented!
+    #     # a <-> b
+    #     graph = {'a': {'b'}, 'b': {'a'}}
+    #     self.assertTrue(Has_cycle(graph))
+
+    def test_no_cycle(self):
+        # a -> b
+        graph = {'a': {'b'}}
+        self.assertFalse(Has_cycle(graph))
+
+    def test_diamon_is_acyclic(self):
+        # a -> b+c -> d
+        DAG = {'a': {'b','c'}, 'b': {'d'}, 'c': {'d'}}
+        self.assertFalse(Has_cycle(DAG))
+
+
+class SelfTest(unittest.TestCase):
+
+    def test_expansions(self):
+        """Tests that expandable commands are not package names"""
+        for exp in expansions.keys():
+            self.assertTrue(exp not in dependencies)
+
+    def test_dependencies_is_acyclic(self):
+        self.assertFalse(Has_cycle(dependencies))
+
+    def test_expansions_is_acyclic(self):
+        self.assertFalse(Has_cycle(expansions))
+
+    def test_all_dependencies_are_packages(self):
+        all_package_names = {case.package_name for case in all_test_cases}
+        for _, deps in dependencies.items():
+            for dep in deps:
+                self.assertTrue(dep in all_package_names)
+
+
+if __name__ == '__main__':
+    test_loader = unittest.TestLoader()
+
+    # Do self test first to fail fast
+    suite = unittest.TestSuite([
+        test_loader.loadTestsFromTestCase(DAG_Algorithm_Test),
+        test_loader.loadTestsFromTestCase(SelfTest)
+    ])
+    result = unittest.TextTestRunner(verbosity=2).run(suite)
+    if not result.wasSuccessful():
+        sys.exit(False)
+
+    commands = sys.argv[1:]
+    sys.argv = [sys.argv[0]] # unittest needs this
+
+    commands.remove('launch')
+    commands.remove('jenkins')
+
+    upstream = 'OFF'
+    if '--upstream' in commands:
```

Comment:
this flag make sense only for cosmo, if we set it for the others like int2lm it will just take the upstream and not cmpile anything. Perhaps we could make it only compatible with cosmo command ?

Reply:
I created and issue https://github.com/C2SM/spack-c2sm/issues/334 for this so we can address it in an other PR, because I think even if the functionality of --upstream is not perfect yet it's not a trap and the PR still improves master.